Check the request format before setting the redirection after the login page

yaacov opened the pull request:
**Description**
On rare events when an ajax request triggers the session timeout, and loads the
login page, the startup page will be a `page not found` error.

This PR prevent the `page not found` error in this cases. In the cases of an ajax request that triggers the timeout, we will not set the `session[:start_url]` to the url of the ajax call.

**Sceenshot**
![screenshot-2016-08-23-19-11-56](https://cloud.githubusercontent.com/assets/2181522/17899894/9be838fa-6965-11e6-8a6a-7b89bf937a57.png)

**Bugzilla**
https://bugzilla.redhat.com/show_bug.cgi?id=1341664
https://bugzilla.redhat.com/show_bug.cgi?id=1370202

yaacov commented:
@miq-bot add_label providers/containers, bug

@simon3z @zeari @alongoldboim please review

yaacov commented:
**A Note about the foreman specs**
In `spec/controllers/provider_foreman_controller_spec.rb` we alter two tests.
This tests where false positive, they got response status 200 OK, but it was from the login page and not from a valid response for the test query.
After this PR any POST request without a valid credentials will get response status 403 Forbidden.

simon3z commented:
@yaacov why are we touching foreman specs?

yaacov commented:
@simon3z 

> why are we touching foreman specs?

The foreman specs are checking that the http response is 200-OK.
The request did not do what they intended to do, but they where successful because they got 200-OK from the login page ( they where redirected to it when authentication failed )
 
After this PR, ajax POST is no longer redirected to the login page, but return an "unauthorized" Error.
To fix the tests you have to login before running the ajax, now they actually work and return 200-OK.

yaacov commented:
@miq-bot add_label darga/yes

https://bugzilla.redhat.com/show_bug.cgi?id=1370202

cc @moolitayer @dclarizio

zeari commented:
LGTM

moolitayer commented:
@yaacov this fix does not resolve the described issue. The unauthorized response causes the page to be stuck with the revolver and manually going to the login page still redirects to the same url.

@martinpovolny @himdel Could you please provide feedback on how to resolve this?

yaacov commented:
@moolitayer :+1: changed the logic of preventing an ajax request from setting `session[:start_url]`
moved from checking for POST , to setting `session[:start_url]` only for requests that has `html` format.

This make more sense since we want to have a `.html` page waiting for us after we login, we do not want to be sent to a `.js` page after we login.

moolitayer commented:
@yaacov this version makes sense to me. What is the behavior now? (redirect? what happens after the login?)

@martinpovolny  @himdel does it makes to not set start_url for ajax|js requests (curious, what are format.js requests?)

yaacov commented:
> hat happens after the login?

@moolitayer when `session[:start_url]` is not set we are redirected to the default startup page. This is not a best solution, but it's beter than redirecting the user to a page not found.

@martinpovolny @himdel is their a way to know what was the last _valid_ url ?

moolitayer commented:
@martinpovolny could you please review

moolitayer commented:
@dclarizio can we get a ui review please

himdel commented:
I'm not sure this makes sense. There are URLs which are perfectly happy to respond both to format.js and to format.html - and rejecting them simply because an ajaxy transition was requested sounds wrong - I think what you really want to check for is whether it was a POST request or a GET request..

himdel commented:
..Also, the reverse is true, format.html does **not** imply you can GET it..

yaacov commented:
@himdel :+1: thanks, changed the logic to GET vs POST

himdel commented:
Thanks, I think this looks good, the code can cope just fine with the value being nil.. Should work :) :+1:

moolitayer commented:
LGTM :+1:

moolitayer commented:
Thanks @himdel

yaacov commented:
This bug is darga/yes, needs back-porting: 
https://bugzilla.redhat.com/show_bug.cgi?id=1370202

simon3z commented:
> This bug is darga/yes, needs back-porting: 
> https://bugzilla.redhat.com/show_bug.cgi?id=1370202

cc @chessbyte

chessbyte commented:
Darga Backport details:
```sh
$ git log

commit 71e564dd576596390c3e4dbb2f9150707f02cc51
Author: Martin Povolny <mpovolny@redhat.com>
Date:   Thu Sep 1 15:59:30 2016 +0200

    Merge pull request #10702 from yaacov/prevent-ajax-calles-from-setting-the-startup-page
    
    Check the request format before setting the redirection after the login page
    (cherry picked from commit d9e7b3e21723687dd12be96084875fdc75e32e88)
```
